src/action.rs

```diff
@@ -122,85 +124,115 @@ impl DynamicUsage for Action<redpallas::Signature<SpendAuth>> {
-            testing::arb_note, TransmittedNoteCiphertext,
+            asset_base::testing::arb_asset_base, commitment::ExtractedNoteCommitment,
+            nullifier::testing::arb_nullifier, testing::arb_note, Note, TransmittedNoteCiphertext,
         },
+        orchard_sighash_versioning::VerSpendAuthSig,
         primitives::redpallas::{
             self,
             testing::{arb_spendauth_signing_key, arb_spendauth_verification_key},
         },
+        primitives::{OrchardDomain, OrchardPrimitives},
         value::{NoteValue, ValueCommitTrapdoor, ValueCommitment},
     };
 
     use super::Action;
 
-    prop_compose! {
-        /// Generate an action without authorization data.
-        pub fn arb_unauthorized_action(spend_value: NoteValue, output_value: NoteValue)(
-            nf in arb_nullifier(),
-            rk in arb_spendauth_verification_key(),
-            note in arb_note(output_value),
-        ) -> Action<()> {
-            let cmx = ExtractedNoteCommitment::from(note.commitment());
-            let cv_net = ValueCommitment::derive(
-                spend_value - output_value,
-                ValueCommitTrapdoor::zero()
-            );
-            // FIXME: make a real one from the note.
-            let encrypted_note = TransmittedNoteCiphertext {
-                epk_bytes: [0u8; 32],
-                enc_ciphertext: [0u8; 580],
-                out_ciphertext: [0u8; 80]
-            };
-            Action {
-                nf,
-                rk,
-                cmx,
-                encrypted_note,
-                cv_net,
-                authorization: ()
+    /// `ActionArb` adapts `arb_...` functions for both Vanilla and ZSA Orchard protocol flavors
+    /// in property-based testing, addressing proptest crate limitations.
+    #[derive(Debug)]
+    pub struct ActionArb<P: OrchardPrimitives> {
+        phantom: core::marker::PhantomData<P>,
+    }
+
+    impl<P: OrchardPrimitives> ActionArb<P> {
+        fn encrypt_note<R: RngCore>(
+            note: Note,
+            memo: Vec<u8>,
+            cmx: &ExtractedNoteCommitment,
+            cv_net: &ValueCommitment,
+            rng: &mut R,
+        ) -> TransmittedNoteCiphertext<P> {
+            let encryptor =
+                NoteEncryption::<OrchardDomain<P>>::new(None, note, memo.try_into().unwrap());
+
+            TransmittedNoteCiphertext {
+                epk_bytes: encryptor.epk().to_bytes().0,
+                enc_ciphertext: encryptor.encrypt_note_plaintext(),
+                out_ciphertext: encryptor.encrypt_outgoing_plaintext(cv_net, cmx, rng),
+            }
+        }
+
+        prop_compose! {
+            /// Generate an action without authorization data.
+            pub fn arb_unauthorized_action(spend_value: NoteValue, output_value: NoteValue)(
+                nf in arb_nullifier(),
+                rk in arb_spendauth_verification_key(),
+                note in arb_note(output_value),
+                asset in arb_asset_base(),
```

Comment:
This is in the wrong place. The asset for the action should be an argument of `arb_unauthorized_action` alongside the spend and output values, otherwise the caller cannot leverage those existing arguments in any meaningful way.
```suggestion
```